Document product and engineering planning process

msennett22 opened the pull request:
This creates a new section in the handbook, `product-engineering`, where we can document things that impact product and engineering equally (like planning), and then it adds documentation for our new planning process. The plan is to move the current `product` and `engineering` folders under this new folder, but that will not happen in this PR.

This PR includes and obsoletes the content in [this doc](https://docs.google.com/document/d/1vYMH3s7B8mP2jRx5_9qYOiVJLs2ZX6T1i9NFHM1xgGk/edit#).

msennett22 committed:
Create Product/Engineering folder

msennett22 committed:
squash! Prettier

nicksnyder committed:
Update from live share with Michal

nicksnyder committed:
updates

nicksnyder committed:
fix links

nicksnyder committed:
fix link

nicksnyder committed:
status definitions

nicksnyder committed:
minor updates

nicksnyder committed:
update link

jasonyavorsky commented:
I love the idea of combining product & engineering under one heirarchy!

jasonyavorsky commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,140 @@
+| Week 12 of current quarter<br>1 week until next quarter | End of quarter retrospective<li>VPP/VPE & Product/Eng Directors retro previous quarter<li>VPP/VPE and CTO team retro previous quarter & finalize OKR scorecard |
+| Week 1 of new quarter | Kick off new quarter<li>VPP/VPE share previous quarter OKR scorecard to Product/Eng Department <li>Communicate department OKRs for upcoming quarter at Company kick off |
+
+## Sharing progress
```
There's some content at https://handbook.sourcegraph.com/product/product_management#okr-and-roadmap-tracking that could either:

- Be linked to from here
- Move here (and then link to here from there)

jasonyavorsky replied:
Not sure which is best.

jasonyavorsky commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,140 @@
+| Week 8 of current quarter<br>4 weeks until next quarter | Draft department level OKRs <li>VPP/VPE & Product/Eng Directors create first draft of OKRs during [VP team sync](team.md#vp-team-sync)<li>VPP/VPE share with exec team and integrate feedback<li>Product/Eng Directors share with their orgs and integrate feedback |
+| Week 9 of current quarter<br>3 weeks until next quarter | Revise department level OKRs<li>VPP/VPE & Product/Eng Directors discuss and revise OKRs during [VP team sync](team.md#vp-team-sync)<li>VPP/VPE share updates with exec team and integrate feedback<li>Product/Eng Directors share updates with their orgs and integrate feedback |
+| Week 10 of current quarter<br>2 weeks until next quarter | Finalize OKRs<li>VPP/VPE & Product/Eng Directors align on final OKRs during [VP team sync](team.md#vp-team-sync)<li>VPP/VPE get exec team buy in to finalize and approve OKRs<li>OKRs added to the handbook & communicated on Slack |
+| Week 12 of current quarter<br>1 week until next quarter | End of quarter retrospective<li>VPP/VPE & Product/Eng Directors retro previous quarter<li>VPP/VPE and CTO team retro previous quarter & finalize OKR scorecard |
```
Around this point in time I think that updating https://handbook.sourcegraph.com/company/strategy should be an official tracked step. What do you think?

```suggestion
| Week 12 of current quarter<br>1 week until next quarter | End of quarter retrospective<li>VPP/VPE & Product/Eng Directors retro previous quarter<li>VPP/VPE and CTO team retro previous quarter & finalize OKR scorecard <li>Update [company strategy and sub-pages](../../company/strategy) |
```

msennett22 replied:
@jyavorska In week 10 we have a step to add the OKRs to the handbook and communicate them on Slack. We can link to the pages as you suggested to ensure people know exactly where in the handbook it will be updated.

jasonyavorsky replied:
I missed that, thanks @msennett22 ! I'll add a suggestion up there instead.

jasonyavorsky commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,140 @@
+- [FY22 Q4 (current week)](https://docs.google.com/presentation/d/1DgY3k684Jn3diCe4GPPcrGt9iaD9-vyndiiJwEoELyE/edit#slide=id.gfca47eca53_0_184)
+- [FY22 Q4 (week over week history)](https://docs.google.com/presentation/d/1d36d47VFSZzQWIhdkZt3JZ3pNUk766yy8iwkFfVpTyY/edit)
+
+### Monthly team updates
```
Adding a weekly section to go with monthly and quarterly so you can't miss it.

```suggestion
### Weekly team updates

These are due each Friday by 19:00 UTC.

- Update the [GitHub project tracker](#github-project-tracker)
- Update the [OKR slides](#okr-slides)

### Monthly team updates
```

Commit:
Remove duplicate content from product handbook

Commit:
Remove duplication from product planning page

Commit:
Link from tracking issues

jasonyavorsky commented:
I made a few small commits to change things in the product handbook that were duplicated to be links to the new content.

Commit:
Fix link

Commit:
Add link to product & engineering direction

jasonyavorsky commented on content/product-engineering/index.md:
```diff
@@ -0,0 +1,16 @@
+
+## Purpose
+
+The role of Product and Engineering is to create value for our customers by delivering a high quality product that solves their problems.
```
```suggestion
The roles of [Product](../product/index.md) and [Engineering](../engineering/index.md) are to create value for our customers by delivering a high quality product that solves their problems.
```

marybelzer commented:
Thoughts on moving the "Product" and Engineering" directories into the "Product-Engineering" directory? Having all three at the same level of the Handbook may make it a bit confusing to navigate and to know where certain things should live

jasonyavorsky commented:
> 
> 
> Thoughts on moving the "Product" and Engineering" directories into the "Product-Engineering" directory? Having all three at the same level of the Handbook may make it a bit confusing to navigate and to know where certain things should live

@marybelzer that's listed as a todo in the description, but as a follow-up PR. Once this merges I'd be happy to take a shot at it, I've built up some tricks for handling large moves like that.

nicksnyder commented:
Yes, we will do that, just not in this particular change so it is easier to review the new content in this change.

msennett22 committed:
Added product/eng FY22Q4 direction content 

Moved the content from https://docs.google.com/document/d/1j_GuBOLXDVpzBzfCWJl-hz2rvqT--IPo3GYU07aG8HE/edit?disco=AAAARZ_6mC4 and redacted customer names and other sensitive info

TODO update FY23 growth target once finalized on company level

msennett22 committed:
squash! Prettier

msennett22 committed:
Fixed link to OKRs

msennett22 commented:
I added the product/eng Q4 direction info and redacted customer/sensitive data. @nicksnyder please let me know if there is anything else we need to remove from this section due to sensitivity

yinquanteo commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,140 @@
+- [FY22 Q4 (current week)](https://docs.google.com/presentation/d/1DgY3k684Jn3diCe4GPPcrGt9iaD9-vyndiiJwEoELyE/edit#slide=id.gfca47eca53_0_184)
+- [FY22 Q4 (week over week history)](https://docs.google.com/presentation/d/1d36d47VFSZzQWIhdkZt3JZ3pNUk766yy8iwkFfVpTyY/edit)
+
+### Monthly team updates
```
@nicksnyder is this something you want from all engineering teams?

I've gone back to my team to gather some initial feedback and overall group sentiment leans towards preferring Code Graph's current GitHub issues approach to providing team updates. Things people liked – process feels more lightweight, everything being in GitHub, better integration with GitHub (referencing issues/PRs). One person did point out they liked the email approach because it had a wider audience.

Regardless of the decision here, I'd love to minimize the need for duplicate updates.

@lguychard @scalabilitysolved @jjeffwarner @felixfbecker @chrispine tagging you here if wanted to comment directly

nicksnyder replied:
Yes the proposal documented in this PR is for each team to send a monthly email (instead of a monthly post to a GitHub issue for that team). The reason for email is that it is easier for stakeholders outside of engineering to subscribe to (only need to subscribe to a single mailing this rather than a bunch of issues), but we could try to automate that to bridge the gap if we need to. If we used one long lived issue per team, eventually they would get really long and hard to navigate.

The goal is for this not to be duplicative with weekly tactical updates on KRs. The audience is different, level of detail is different (due to monthly vs weekly cadence)

Definitely open to feedback here!

jasonyavorsky reviewed (approved):
Left a few ideas. Looks good!

felixfbecker commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,140 @@
+| Week 3          | Gather input from stakeholders<li>VPP/VPE get feedback from exec team and key stakeholders                                                |
+| Week 4          | Update annual vision <li>VPP/VPE update annual vision draft based on feedback                                                             |
+| Week 5          | Share & gather feedback <li>VPP/VPE share annual vision with Product/Eng Directors and gather feedback                                    |
+| Week 6          | Finalize annual vision document <li>VPs update and finalize based on feedback <li>Annual vision added to handbook & communicated on Slack |
```
**issue**: The `<li>` (list item) tags in these tables are missing enclosing `<ul>` `</ul>` (unordered list) tags and closing `</li>` tags, which causes rendering issues, e.g. in Firefox:

<img width="801" alt="image" src="https://user-images.githubusercontent.com/10532611/141464075-4262fc59-fbf4-43d2-b57c-a35570876817.png">

```suggestion
| Week 6          | Finalize annual vision document <ul> <li>VPs update and finalize based on feedback</li> <li>Annual vision added to handbook & communicated on Slack</li> </ul> |
```

(needs to be done for all `<li>`s in this PR)

msennett22 committed:
Update content/product-engineering/planning.md

The <li> (list item) tags in these tables are missing enclosing <ul> </ul> (unordered list) tags and closing </li> tags, which causes rendering issues, e.g. in Firefox:

Co-authored-by: Felix Becker <felix.b@outlook.com>

msennett22 committed:
squash! Prettier

jasonyavorsky commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,140 @@
+| Week 7 of current quarter<br>5 weeks until next quarter | Gather input from stakeholders on upcoming priorities <li>VPP/VPE gather input from exec team on important problems to solve next quarter <li>Product/Eng Directors gather input from teams on next quarter priorities |
+| Week 8 of current quarter<br>4 weeks until next quarter | Draft department level OKRs <li>VPP/VPE & Product/Eng Directors create first draft of OKRs during [VP team sync](team.md#vp-team-sync)<li>VPP/VPE share with exec team and integrate feedback<li>Product/Eng Directors share with their orgs and integrate feedback |
+| Week 9 of current quarter<br>3 weeks until next quarter | Revise department level OKRs<li>VPP/VPE & Product/Eng Directors discuss and revise OKRs during [VP team sync](team.md#vp-team-sync)<li>VPP/VPE share updates with exec team and integrate feedback<li>Product/Eng Directors share updates with their orgs and integrate feedback |
+| Week 10 of current quarter<br>2 weeks until next quarter | Finalize OKRs<li>VPP/VPE & Product/Eng Directors align on final OKRs during [VP team sync](team.md#vp-team-sync)<li>VPP/VPE get exec team buy in to finalize and approve OKRs<li>OKRs added to the handbook & communicated on Slack |
```
```suggestion
| Week 10 of current quarter<br>2 weeks until next quarter | Finalize OKRs<li>VPP/VPE & Product/Eng Directors align on final OKRs during [VP team sync](team.md#vp-team-sync)<li>VPP/VPE get exec team buy in to finalize and approve OKRs<li>OKRs added to the handbook and [company strategy and sub-pages](../../company/strategy) updated & communicated on Slack |
```

Commit:
Update content/product-engineering/planning.md

Commit:
Fix link

msennett22 committed:
Create planning.md

Updated the <li> tags in these tables to include enclosing <ul> </ul> tags and closing </li> tags to fix rendering issues, e.g. in Firefox

msennett22 committed:
squash! Prettier

Commit:
Update content/product-engineering/planning.md

Commit:
Update content/product-engineering/index.md

Commit:
Merge branch 'main' into product-engineering-planning

nicksnyder commented on content/product-engineering/direction.md:
```diff
@@ -0,0 +1,17 @@
+# Product and engineering direction
+
+Sourcegraph is a product that developers love, and once we land inside an organization usage grows organically! To set us up to grow our business by ~X (TODO update when finalized) in FY23 we need to repeatably drive more users to the point where they are using Sourcegraph on their private code with their team. To achieve this outcome, we are going to pursue three strategic objectives in FQ4.
```
```suggestion
Sourcegraph is a product that developers love, and once we land inside an organization usage grows organically! To set us up to grow our business in FY23 we need to repeatably drive more users to the point where they are using Sourcegraph on their private code with their team. To achieve this outcome, we are going to pursue three strategic objectives in FQ4.
```

nicksnyder committed:
Update content/product-engineering/direction.md

jasonyavorsky commented on content/company/strategy/index.md:
```diff
@@ -115,14 +115,16 @@ Coding itself is a technology, and we anticipate the technology adoption curve w
 Within this overall product strategy, individual product teams set their own goals and roadmaps aligned to the [company OKRs](../goals/index.md), and each have a strategy page with all the details of what they are working on next and why. There is also an [(currently internal only) presentation](https://docs.google.com/presentation/d/1o3R8WUIhzzRz0x5laTwVcizOzVWrMBe5MCAz74H45Ss/edit#slide=id.gd8d ...
 
-### [Code Graph](code-graph/index.md)
+### [Product & Engineering](../../product-engineering/direction.md)
```
Propose renaming this to `strategy.md` to be consistent with what we're calling other strategy pages (cc @nicksnyder @serinadawn )

nicksnyder replied:
done

msennett22 commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,147 @@
+# Product and engineering planning
+
```
Proposing to add a section about our philosophy/principles. It could potentially look something like this. 

##Planning philosophy and principles 
Our planning process defines how we meet and execute to create our annual and quarterly goals. Our philosophy is based on the following principles:

- Clear direction: by setting clear direction on the department level we are able to stay focused throughout the year and have appropriate trade-off discussions if necessary. 

- Coupling insights: our teams are the domain experts and their valuable insights coupled with the overall company strategy are critical to both enable and drive our strategic planning.

- Creating impact : to set ourselves up for success we want to understand and be able to measure the impact and value the work we commit to will have on customers, our internal teams, and the overall success of Sourcegraph.

jasonyavorsky replied:
@msennett22 FYI there is a philosophy section at `/product-engineering/planning#philosophy` (could be brand new, not sure) and it touches on some of these.

nicksnyder replied:
Yes, just added that based on feedback yesterday and this comment. Going to resolve this thread.

nicksnyder committed:
direction -> strategy

nicksnyder committed:
link to company strategy

nicksnyder committed:
fix link

nicksnyder committed:
some philosophy

Commit:
Bring glossary from product handbook

Commit:
Fix link

Commit:
Don't have two different pages describing planning steps

jasonyavorsky commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,183 @@
+# Product and engineering planning
+
+- [Product and engineering planning](#product-and-engineering-planning)
```
Do we need both a manually written table of contents in addition to the automatically generated one to the right?

nicksnyder replied:
I found this one helpful (and it is auto-generated by me via extension), but I will remove it. The index to the right in the handbook is missing two things IMO:
1. I think it should include H2s under H1s with indent.
1. Admittedly personal preference, but I think outlines should be on the left. On the right feels too hidden.

Issue filed here: https://github.com/sourcegraph/handbook/issues/901

jasonyavorsky replied:
I actually like them on the left too, and I don't like the way it hides h2s today. At the same time it does feel strange to see two copies of the table of contents side by side when you go to the page. Maybe there's an easy fix for the official table of contents.

Commit:
Remove duplicated content

Commit:
Fix indent

jasonyavorsky commented:
This morning I made it so there isn't also a planning page under product management, instead it just links to the product-engineering planning page.

jasonyavorsky commented on content/product-engineering/direction.md:
```diff
@@ -0,0 +1,17 @@
+# Product and engineering direction
```
Maybe rename this file to strategy.md? Not super important but it's more consistent.

dan-mckean commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,214 @@
+
+To acheive a good plan, our planning process based on the following principles:
+
+- Everyone on our team has context and insights that are valuable to inputs into the planning process.
```
```suggestion
- Everyone on our team has context and insights that are valuable inputs into the planning process.
```

dan-mckean commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,214 @@
+To acheive a good plan, our planning process based on the following principles:
+
+- Everyone on our team has context and insights that are valuable to inputs into the planning process.
+- success we want to understand and be able to measure the impact and value the work we commit to will have on customers, our internal teams, and the overall success of Sourcegraph.
```
The start of this makes no sense and it's not clear what it was meant to say.

nicksnyder replied:
This whole section is still wip. Will be working on this more today.

dan-mckean commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,214 @@
+
+- Everyone on our team has context and insights that are valuable to inputs into the planning process.
+- success we want to understand and be able to measure the impact and value the work we commit to will have on customers, our internal teams, and the overall success of Sourcegraph.
+- Arriving at a good plan is an iterative process that is neither exclusively top-down nor exclusively bottom-up because there is a cyclical information dependency. Company level strategy is informed by what is working and not working about our business (context our teams have), and team plans are informed by what our company level strategy is (a decision ultimately owned by our CEO).
```
Love this line ❤️

dan-mckean commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,214 @@
+- Attendees
+  - VPP, VPE, and their direct reports
+- Prework
+  - DRI responsible for each KR fills out their retrospective slide in preparation for meeting, colleting any necessary information from their team in advance.
```
I feel like this line—specifically the bit about collecting information from teams—belies a lot of work, and it would be worth somehow making this more explicit to ensure it happens properly, far enough in advance of the VP/Director meeting.

nicksnyder replied:
What work needs to happen is an implementation detail for the directors to figure out. This is just documenting the API.

dan-mckean commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,214 @@
+    - What didn’t go well (struggles and challenges) for yourself and for your team?
+    - What did you and/or your team learn?
+- Deliverable
+  - Finalized OKR scorecard deck ready to be shared with the company
```
What happens next? Do you share the learnings and give space for others outside of this small working group to suggest solutions?

nicksnyder replied:
I think we will flesh this section out when we get close to doing the retro for Q4

dan-mckean commented on content/product-engineering/team.md:
```diff
@@ -0,0 +1,36 @@
+- [Product org chart](../product/product_org.md)
+- [Engineering org chart](../engineering/eng_org.md)
+
+## CTO team sync
```
What is the CTO team? Worth being more explicit!

jasonyavorsky commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,214 @@
+
+- Strategy: Our [strategy](strategy.md) is a concise description of how product and engineering contribute to the top-level company objectives. It sets a clear direction so we stay focused on what is important and have appropriate trade-off discussions if necessary.
+- OKRs: [OKRs](../company/goals/guidelines.md) define how we measure success.
+- Roadmap items: The concrete high-level things that we are delivering to achieve success (as measured by OKRs).
```
Maybe add milestones as a fourth definition here because we're now tracking those as separate entities?

felixfbecker commented on content/product-engineering/planning.md:
```diff
@@ -0,0 +1,214 @@
+
+- The **Product and Engineering** tab tracks the results that the product and engineering VPs and directors review every week at the [VP team sync](#vp-team-sync).
+- Each org (Cloud, Enablement, Code graph) has a tab that tracks the results for that org. Org directors review this dashboard each week with their team.
+- The **Engineering** tab tracks results that are relevant only to engineering or engineering leadership. This is reviewed every week in the [VPE team sync](#todo-link).
```
Leftover TODO link here

@jyavorska  shouldn't the link checker have caught this? 🤔

jasonyavorsky replied:
@felixfbecker I removed the link checker I made because it didn't work properly. The only way to really make it work consistently is post-build and export.

Commit:
Merge branch 'main' into product-engineering-planning

Commit:
Merge branch 'main' into product-engineering-planning

nicksnyder committed:
remove index

nicksnyder committed:
fix sentence

nicksnyder committed:
address comments

nicksnyder committed:
fix link

nicksnyder commented:
@christinaforney and I aligned on what we want out of the tracker today but I ran out of time to make any updates to this PR. I will keep working on this PR tomorrow and the rest of this week to get this merged ASAP. 

Here is the short version...

There are two kinds of line items: KRs and Roadmap items.
* KRs are the implementation-independent definition of success. Generally, these are based on metrics.
* Roadmap items are concrete things the team is delivering to achieve the definition of success. Generally, these are feature/product launches and each roadmap items has a "Target date" that the implementing team is responsible for setting and updating for accuracy.
  * For most rows, the target date just needs to be a best effort (conservative) estimate of when we will be done. Changes to the target date are ok, just log what changed and why in the linked issue. This helps me/Christina/CE/Sales/Marketing generally know when things are going to be available so we can plan ahead. The more accurate dates that you can provide, the quicker we will be able to get customers using what you have released on day 1 so you aren't waiting for feedback.
  * For a small number of rows that are our biggest launches and we want to do a big marketing effort around, we will need to have a more firm date that we can commit to (for embargo reasons) when we get closer to launch. We will ensure that the people responsible for these deliverables know what the expectations are.

Tracker tabs:
* Department KRs tab unchanged. It only contains top-level product/eng KRs. This is what we review on a weekly basis with product/eng directors and what we are reporting on to the rest of the business and exec team.
* Org KRs new tab. This has per-org level KRs. The expectation is that product/eng directors are reviewing these weekly with relevant teams. Me/Christina will deep-dive review these with product/eng directors monthly, and spot check as necessary or whenever something is escalated.
* Team deliverables new tab. This has top-level roadmap items per team so Christina and I know what the big picture of each team is working on. If when a team doesn't have a roadmap item, then we will include the KR as a placeholder, but once the team has identified at least one roadmap item, we will remove the KR from this view. Teams that have line of sight into multiple future deliverables are encouraged to include those, but don't artificially flesh out a roadmap that you don't have confidence in. 
* Per-org level tabs are owned by the directors for each org. Christina and I don't care about what is in those tabs or how they are used.

Christina and I also recorded a video running through the dashboard, but zoom is not done processing that yet....I'll post it in [#product-engineering-planning](https://sourcegraph.slack.com/archives/C02LHQW7RC0/p1637104051148100) when it is ready.


Would love feedback on this latest iteration of the proposal.

jasonyavorsky commented:
> There are two kinds of line items: KRs and Roadmap items.

Does this mean the milestone issues we just added are going away?

christinaforney commented:
> Does this mean the milestone issues we just added are going away?

No, we migrated all Milestones to be Roadmap items. The difference is the view level - if it shows up at Prod/Eng (Department) level vs. Org level.

nicksnyder committed:
updates

nicksnyder committed:
fix links

nicksnyder committed:
fix link and add back strategy page content

nicksnyder committed:
Merge branch 'main' into product-engineering-planning

nicksnyder commented:
I am happy with where this is at so going to bias toward merging now and iterating on the content.

@jyavorska I removed some of the content that you added to the new planning page. I really want this page to be focused on the process itself and not the details of how to prioritize things.

jasonyavorsky commented:
Sure @nicksnyder sorry those changes weren't more helpful. We can find a different way to disambiguate the two planning process pages in a later follow up.

jasonyavorsky commented:
Done: https://github.com/sourcegraph/handbook/pull/948
